src/Raku/ast/code.rakumod

```diff
@@ -1826,6 +1827,43 @@ class RakuAST::Sub
             self.IMPL-WRAP-SCOPE-HANDLER-QAST($context,
                 self.IMPL-APPEND-SIGNATURE-RETURN($context, $!body.IMPL-TO-QAST($context))))
     }
+
+    method PERFORM-CHECK(Resolver $resolver, RakuAST::IMPL::QASTContext $context) {
+        return Nil unless self.multiness eq 'multi';
+
+        my $proto := $resolver.resolve-lexical('&' ~ self.name.canonicalize, :current-scope-only).compile-time-value.derive_dispatcher;
```

Comment:
This looks wrong. PERFORM-CHECK is usually called after we are done parsing. That means the resolver will have exited all lexical scopes already. E.g.
```
{
     multi sub foo() {}
     multi sub foo() {} # we're checking this
} 
# <- resolver will look from here
```

The only place where the resolver is guaranteed to be in the right state for resolving lexicals is the attach method.

Reply:
But we are dealing with a (likely) generated proto declaration and that happens outside of attach? 

We also need all of the begin time setup to be performed (so for instance 'is default' has been mixed in).

Is it okay to store a reference to the scope in attach and use that reference in PERFORM-CHECK?

Reply:
Well, sins of the past, what can I say?
While in BEGIN processing we're no longer guaranteed that the resolver is in the correct state, it's still likely as BEGIN will be handled e.g. when we encounter the closing } of a routine. Which will be good enough to find the proto. When using the parser that is. For synthetically generated ASTs it may be quite different and it's likely that we need to re-visit the existing code as well.

That said, there should not be a reason for you to look up the proto in the first place. After all you already have access to one of the multi candidates. A multi candidate knows its proto. `self.meta-object.dispatcher` will give it to you. Also there's no need to derive a dispatcher here. You don't want to add any candidates. You're just checking the existing ones.

Reply:
Ah, yes, I should have remembered that I can get the dispatcher directly from the mega object.

So doing it that way, I assume there is no issue with the check remaining at check time?

Reply:
Not at all. Check time is perfect for this.

Reply:
This is addressed in the newest implementation. As per the discussion in #5455, it sounds like we are going with a sorry for this case.

Reply:
Since there are valid reasons for having multiple candidates matching, as @niner described, shouldn't this be a worry instead then?

Reply:
As far as I see there are three options:
- We allow identically named methods. I oppose this. This feature hasn't been explored at all yet. The potential for confusion is large. (You don't notice you have two identical methods, only look at one of the two and wonder what's happening.)
- We worry this. I also oppose this, but slightly less. People will be warned, but we still open the door for such code to appear in the Raku ecosystem. And then we're locked in to support this.
- We sorry this. Should someone want this later on and the use case is legit, we can then start evaluating the pros and cons and take a well thought through decision to add it or not.